lib/message_request.js

```diff
@@ -0,0 +1,166 @@
+
+  if (!this._isValid()) {
+    throw new Error('Invalid request message.');
+  }
+};
+
+Request.buildGet = function (scope, options) {
+  return new Request('get', scope).setOptions(options);
+};
+
+Request.buildPublish = function (scope, value) {
+  var request = new Request('publish', scope);
+  request.setAttr('value', value);
+  return request;
+};
+
+Request.buildPush = function (scope, resource, action, value) {
+  var request = new Request('push', scope);
+  request.setAttr('resource', resource);
+  request.setAttr('action', action);
+  request.setAttr('value', value);
+
+  return request;
+};
+
+Request.buildNameSync = function (scope, options) {
+  return new Request('nameSync', scope).setOptions(options);
+};
+
+Request.buildSet = function (scope, value, clientData, key, userType) {
+  var request = new Request('set', scope);
+  request.setAttr('value', value);
+  request.setAttrOptional('key', key);
+  request.setAttrOptional('type', userType);
+  if (typeof(clientData) != 'function') {
+    request.setAttrOptional('clientData', clientData);
+  }
+  return request;
+};
+
+Request.buildSync = function (scope, options) {
+  return new Request('sync', scope).setOptions(options);
+};
+
+Request.buildSubscribe = function (scope, options) {
+  return new Request('subscribe', scope).setOptions(options);
+};
+
+Request.buildUnsubscribe = function (scope, options) {
+  return new Request('unsubscribe', scope);
+};
+
+// TO DO: config class should return the required data via getters
+Request.prototype.setAuthData = function (configuration) {
+  if (configuration && configuration.auth) {
+    this.setAttr('auth', configuration.auth);
+    this.setAttr('userId', configuration.userId);
+    this.setAttr('userType', configuration.userType);
+    this.setAttr('accountName', configuration.accountName);
+  }
+};
+
+Request.getAttr = function (message, attr) {
+  return (message && message[attr]);
+};
+
+// Instance methods
+
+Request.prototype.getMessage = function () {
+  return this.message;
+};
+
+Request.prototype.setOptions = function (options) {
+  if (options && typeof options != 'function') {
+    this.message.options = options;
+    this.version = 2;
+  } else {
+    this.version = 1;
+  }
+
+  return this;
+};
+
+Request.prototype.isPresence = function () {
+  return this.type === 'presence';
+};
+
+Request.prototype.setAttrOptional = function (keyName, keyValue) {
+  if (keyName && (keyValue === 0 || keyValue)) {
+    this.message[keyName] = keyValue;
+  }
+};
+
+Request.prototype.setAttr = function (keyName, keyValue) {
+  //if (!keyName || !keyValue) {
+  if (!(keyName && (keyValue === 0 || keyValue))) {
+    throw new Error('Invalid request attribute');
+  }
+  this.message[keyName] = keyValue;
+};
+
+Request.prototype.setAttrOptional = function (keyName, keyValue) {
+  if (keyName && (keyValue === 0 || keyValue)) {
+    this.message[keyName] = keyValue;
+  }
+};
+
+Request.prototype.getAttr = function (keyName) {
+  if (keyName) {
```

Comment:
For optional keys (see buildSet() above), yes, keys can be null.  As for checks like (keyValue === 0), this is a valid value for userType, but javascript interprets a value of 0 as falsy.  I could just check for a non-null keyName and let keyValue be any value at all, but we would end up with missing scopes, options, etc. Push those checks into radar_client.js is not something I want to do, when we can enforce key/val requirements in one place.